integrations/redis.go

```diff
@@ -0,0 +1,154 @@
+// and metrics.
+type RedisBaseplateHook struct {
+	clientName string
+	metrics    metricsbp.Statsd
+}
+
+var _ redis.Hook = RedisBaseplateHook{}
+
+// NewRedisBaseplateHook creates a new RedisBaseplateHook to integrate a Redis
+// client from `go-redis` with Baseplate monitoring and tracing.
+func NewRedisBaseplateHook(ctx context.Context, clientName string, metricsConfig metricsbp.StatsdConfig) RedisBaseplateHook {
+	metrics := metricsbp.NewStatsd(ctx, metricsConfig)
+	return RedisBaseplateHook{clientName: clientName, metrics: metrics}
+}
+
+// BeforeProcess starts a client Span before processing a Redis command and
+// starts a timer to record how long the command took.
+func (h RedisBaseplateHook) BeforeProcess(ctx context.Context, cmd redis.Cmder) (context.Context, error) {
+	return h.startChildSpan(ctx, cmd.Name()), nil
+}
+
+// AfterProcess ends the client Span started by BeforeProcess, publishes the
+// time the Redis command took to complete, and a metric indicating whether the
+// command was a "success" or "fail"
+func (h RedisBaseplateHook) AfterProcess(ctx context.Context, cmd redis.Cmder) error {
+	return h.endChildSpan(ctx, cmd.Err())
+}
+
+// BeforeProcessPipeline starts a client span before processing a Redis pipeline
+// and starts a timer to record how long the pipeline took.
+func (h RedisBaseplateHook) BeforeProcessPipeline(ctx context.Context, cmds []redis.Cmder) (context.Context, error) {
+	return h.startChildSpan(ctx, "pipeline"), nil
+}
+
+// AfterProcessPipeline ends the client span started by BeforeProcessPipeline,
+// publishes the time the Redis pipeline took to complete, and a metric
+// indicating whether the pipeline was a "success" or "fail"
+func (h RedisBaseplateHook) AfterProcessPipeline(ctx context.Context, cmds []redis.Cmder) error {
+	var err error
+	for _, cmd := range cmds {
+		err = cmd.Err()
+		if err != nil {
+			break
+		}
+	}
+	return h.endChildSpan(ctx, err)
+}
+
+func (h RedisBaseplateHook) startChildSpan(ctx context.Context, cmdName string) context.Context {
+	span := tracing.GetServerSpan(ctx)
+	if span == nil {
+		return ctx
+	}
+	ctx = setRedisCommandStartTime(ctx)
+	ctx = setRedisCommandName(ctx, cmdName)
+	spanName := fmt.Sprintf("%s.%s", h.clientName, cmdName)
+	ctx, _ = span.CreateClientChildForContext(ctx, spanName)
+	return ctx
+}
+
+func (h RedisBaseplateHook) endChildSpan(ctx context.Context, err error) error {
+	span := tracing.GetChildSpan(ctx)
+	if span == nil {
+		return nil
+	}
+	cmd := getRedisCommandName(ctx)
+	if cmd != "" {
+		h.recordTime(ctx, cmd)
+		h.recordStatus(cmd, err)
+	}
+	return span.End(ctx, err)
+}
+
+func (h RedisBaseplateHook) recordTime(ctx context.Context, name string) {
+	start := getRedisCommandStartTime(ctx)
+	// The default value for time.Time has a value of "0", just return if that
+	// is returned rather than comparing to the current time.
+	if start.Nanosecond() == 0 {
+		return
+	}
+
+	duration := time.Now().UTC().Sub(start).Nanoseconds()
```

Comment:
The `.UTC()` part is unnecessary.

go-kit metrics package actually have a timer implementation. Basically in `startChildSpan` instead of storing the `time.Time` in the context, you store a [`metrics.Timer`](https://godoc.org/github.com/go-kit/kit/metrics#NewTimer) instead. Then here, if you get a non-nil timer out of the context, you can just call its `ObserveDuration` here. We can also add our own wrapper to make sure it uses millisecond as the unit in `metricsbp` package.

Reply:
https://github.com/reddit/baseplate.go/pull/40